Constant sequence can not be passed as iterator openArray parameter to countup()

### Description

```
iterator demo(a: openArray[int]): int =
  for k in countUp(a[0], 19):
    yield 23

for k in demo(@[17]):
  echo k
```

### Nim Version

Nim Compiler Version 2.2.0 [Linux: amd64]
Compiled at 2024-10-02
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 78983f1876726a49c69d65629ab433ea1310ece1
active boot switches: -d:release

### Current Output

```text
/home/clean/.cache/nim/bug20241014_d/@mbug20241014.nim.c: In function ‘NimMainModule’:
/home/clean/.cache/nim/bug20241014_d/@mbug20241014.nim.c:138:142: error: lvalue required as unary ‘&’ operand
  138 |                         nimlf_(46, "/home/clean/.choosenim/toolchains/nim-2.2.0/lib/system/iterators_1.nim");                   colontmp_ = (&((NI)17));
      |                                                                                                                                              ^
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread   -I/home/clean/.choosenim/toolchains/nim-2.2.0/lib -I/home/clean/project/bug20241014 -o /home/clean/.cache/nim/bug20241014_d/@mbug20241014.nim.c.o /home/clean/.cache/nim/bug20241014_d/@mbug20241014.nim.c' failed with exit code: 1
```


### Expected Output

```text
None. This should compile and not generate a gcc or clang error.
```


### Known Workarounds

Using `seq[int]` as parameter type instead of `openArray[int]` makes the example compile.

### Additional Information

The problem can not be reproduced using lower versions of nim than version 2.2.0, which includes version 2.0.10.
The current `devel` is also affected, which is

Nim Compiler Version 2.2.1 [Linux: amd64]
Compiled at 2024-10-14
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 07628b0dec43dd417b6b536eff1c0f0a5c539b6d
active boot switches: -d:release

--- a/compiler/transf.nim
+++ b/compiler/transf.nim
@@ -56,7 +56,6 @@ type
     contSyms, breakSyms: seq[PSym]  # to transform 'continue' and 'break'
     deferDetected, tooEarly: bool
     isIntroducingNewLocalVars: bool  # true if we are in `introducingNewLocalVars` (don't transform yields)
-    inAddr: bool
     flags: TransformFlags
     graph: ModuleGraph
     idgen: IdGenerator
@@ -103,12 +102,12 @@ proc newTemp(c: PTransf, typ: PType, info: TLineInfo): PNode =
   else:
     result = newSymNode(r)
 
-proc transform(c: PTransf, n: PNode): PNode
+proc transform(c: PTransf, n: PNode, noConstFold = false): PNode
 
-proc transformSons(c: PTransf, n: PNode): PNode =
+proc transformSons(c: PTransf, n: PNode, noConstFold = false): PNode =
   result = newTransNode(n)
   for i in 0..<n.len:
-    result[i] = transform(c, n[i])
+    result[i] = transform(c, n[i], noConstFold)
 
 proc newAsgnStmt(c: PTransf, kind: TNodeKind, le: PNode, ri: PNode; isFirstWrite: bool): PNode =
   result = newTransNode(kind, ri.info, 2)
@@ -481,8 +480,8 @@ proc transformYield(c: PTransf, n: PNode): PNode =
     result.add(introduceNewLocalVars(c, c.transCon.forLoopBody))
     c.isIntroducingNewLocalVars = false
 
-proc transformAddrDeref(c: PTransf, n: PNode, kinds: TNodeKinds): PNode =
-  result = transformSons(c, n)
+proc transformAddrDeref(c: PTransf, n: PNode, kinds: TNodeKinds, isAddr = false): PNode =
+  result = transformSons(c, n, noConstFold = isAddr)
   # inlining of 'var openarray' iterators; bug #19977
   if n.typ.kind != tyOpenArray and (c.graph.config.backend == backendCpp or sfCompileToCpp in c.module.flags): return
   var n = result
@@ -1010,7 +1009,7 @@ proc transformDerefBlock(c: PTransf, n: PNode): PNode =
     result[i] = e0[i]
   result[e0.len-1] = newTreeIT(nkHiddenDeref, n.info, n.typ, e0[e0.len-1])
 
-proc transform(c: PTransf, n: PNode): PNode =
+proc transform(c: PTransf, n: PNode, noConstFold = false): PNode =
   when false:
     var oldDeferAnchor: PNode
     if n.kind in {nkElifBranch, nkOfBranch, nkExceptBranch, nkElifExpr,
@@ -1077,15 +1076,9 @@ proc transform(c: PTransf, n: PNode): PNode =
   of nkCallKinds:
     result = transformCall(c, n)
   of nkHiddenAddr:
-    let oldInAddr = c.inAddr
-    c.inAddr = true
-    result = transformAddrDeref(c, n, {nkHiddenDeref})
-    c.inAddr = oldInAddr
+    result = transformAddrDeref(c, n, {nkHiddenDeref}, isAddr = true)
   of nkAddr:
-    let oldInAddr = c.inAddr
-    c.inAddr = true
-    result = transformAddrDeref(c, n, {nkDerefExpr, nkHiddenDeref})
-    c.inAddr = oldInAddr
+    result = transformAddrDeref(c, n, {nkDerefExpr, nkHiddenDeref}, isAddr = true)
   of nkDerefExpr:
     result = transformAddrDeref(c, n, {nkAddr, nkHiddenAddr})
   of nkHiddenDeref:
@@ -1165,7 +1158,7 @@ proc transform(c: PTransf, n: PNode): PNode =
   let exprIsPointerCast = n.kind in {nkCast, nkConv, nkHiddenStdConv} and
                           n.typ != nil and
                           n.typ.kind == tyPointer
-  if not exprIsPointerCast and not c.inAddr:
+  if not exprIsPointerCast and not noConstFold:
     var cnst = getConstExpr(c.module, result, c.idgen, c.graph)
     # we inline constants if they are not complex constants:
     if cnst != nil and not dontInlineConstant(n, cnst):

--- a/compiler/transf.nim
+++ b/compiler/transf.nim
@@ -1077,7 +1077,10 @@ proc transform(c: PTransf, n: PNode): PNode =
   of nkCallKinds:
     result = transformCall(c, n)
   of nkHiddenAddr:
+    let oldInAddr = c.inAddr
+    c.inAddr = true
     result = transformAddrDeref(c, n, {nkHiddenDeref})
+    c.inAddr = oldInAddr
   of nkAddr:
     let oldInAddr = c.inAddr
     c.inAddr = true

--- a/tests/iter/tfoldedaddr.nim
+++ b/tests/iter/tfoldedaddr.nim
@@ -24,3 +24,9 @@ block: # issue #24305 with array
 
   for k in demo([17]):
     echo k
+
+block: # related regression
+  proc main =
+    let a = [0, 1, 2]
+    let x = addr a[low(a)]
+  main()

--- a/tests/iter/tfoldedaddr.nim
+++ b/tests/iter/tfoldedaddr.nim
@@ -1,0 +1,26 @@
+discard """
+  output: '''
+23
+23
+23
+23
+23
+23
+'''
+"""
+
+block: # issue #24305
+  iterator demo(a: openArray[int]): int =
+    for k in countUp(a[0], 19):
+      yield 23
+
+  for k in demo(@[17]):
+    echo k
+
+block: # issue #24305 with array
+  iterator demo(a: array[1, int]): int =
+    for k in countUp(a[0], 19):
+      yield 23
+
+  for k in demo([17]):
+    echo k

